File: Multi_routes_deploy/sources/model/CapacityCalculate.cpp
#include "CapacityCalculate.h"

namespace fleetdeployment {


    CapacityCalculate::CapacityCalculate(const InputData& in, const Parameter& p)
        : in_(in), p_(p)
    {
        try {
            env_ = IloEnv();
            model_ = IloModel(env_);
            cplex_ = IloCplex(model_);
            
            cplex_.setOut(env_.getNullStream());
            cplex_.setParam(IloCplex::Param::WorkMem, MaxWorkMem);
            cplex_.setParam(IloCplex::Param::TimeLimit, MIPTimeLimit);
            cplex_.setParam(IloCplex::Param::MIP::Tolerances::MIPGap, MIPGapLimit);
            cplex_.setParam(IloCplex::Param::Threads, MaxThreads);

            // create basic decision and add basic constraints
            SetDecisionVars();
            SetObjectives();
            SetConstraints();

            SolveModel();

        }
        catch (const IloException& e) {
            std::cerr << "Error: " << e.getMessage() << std::endl;
        }
    }

    void CapacityCalculate::SetDecisionVars()
    {
        v_var_.resize(p_.GetVesselRouteSet().size(), std::vector<IloNumVar>(p_.GetVesselPathSet().size()));
        x_var_.resize(p_.GetDemand().size());
        y_var_.resize(p_.GetDemand().size());
        z_var_.resize(p_.GetDemand().size());
        g_var_.resize(p_.GetDemand().size());


        for (size_t w = 0; w < p_.GetVesselPathSet().size(); ++w) {
            int r = in_.GetVesselPathSet()[w].GetRouteID() - 1;
            std::string var_name = "v(" + std::to_string(r + 1) + ")(" + std::to_string(w + 1) + ")";
            v_var_[r][w] = IloNumVar(env_, 0, IloInfinity, var_name.c_str());
        }

        for (size_t i = 0; i < p_.GetDemand().size(); ++i) {
            const Request& od = in_.GetRequests()[i];

            std::vector<IloNumVar> xxxVar_k(od.GetNumberOfLadenPath());
            std::vector<IloNumVar> yyyVar_k(od.GetNumberOfLadenPath());
            std::vector<IloNumVar> zzzVar_k(od.GetNumberOfEmptyPath());

            for (size_t j = 0; j < od.GetNumberOfLadenPath(); ++j) {
                std::string var_name_x = "x(" + std::to_string(i + 1) + ")";
                xxxVar_k[j] = IloNumVar(env_, 0, IloInfinity, var_name_x.c_str());
                std::string var_name_y = "y(" + std::to_string(i + 1) + ")";
                yyyVar_k[j] = IloNumVar(env_, 0, IloInfinity, var_name_y.c_str());
            }
            for (size_t j = 0; j < od.GetNumberOfEmptyPath(); ++j) {
                std::string var_name_z = "z(" + std::to_string(i + 1) + ")";
                zzzVar_k[j] = IloNumVar(env_, 0, IloInfinity, var_name_z.c_str());
            }

            x_var_[i] = std::move(xxxVar_k);
            y_var_[i] = std::move(yyyVar_k);
            z_var_[i] = std::move(zzzVar_k);

            std::string var_name_g = "g(" + std::to_string(i + 1) + ")";
            g_var_[i] = IloNumVar(env_, 0, IloInfinity, var_name_g.c_str());
        }
    }

    void CapacityCalculate::SetObjectives()
    {
        try {

            IloExpr  expr(env_);

            // Item 1: Operating Cost
            for (size_t w = 0; w < p_.GetVesselPathSet().size(); ++w) {
                int r = in_.GetVesselPathSet()[w].GetRouteID() - 1;
                expr += p_.GetShipRouteAndVesselPath()[r][w] * v_var_[r][w];
            }

            // Loop over demands
            for (size_t i = 0; i < p_.GetDemand().size(); ++i) {
                // Item 2: Penalty Cost of unsatisfied Demand
                expr += p_.GetPenaltyCostForDemand()[i] * g_var_[i];

                const Request& od = in_.GetRequests()[i];

                // Laden paths
                for (size_t k = 0; k < od.GetNumberOfLadenPath(); ++k) {
                    int j = od.GetLadenPathIndexes()[k];
                    // Item 3: Demurrage of self-owned and leased containers, and Rental cost on laden paths
                    expr += p_.GetLadenPathDemurrageCost()[j] * x_var_[i][k];
                    expr += p_.GetLadenPathDemurrageCost()[j] * y_var_[i][k];
                    expr += p_.GetRentalCost() * p_.GetTravelTimeOnPath()[j] * y_var_[i][k];
                }

                // Empty paths
                for (size_t k = 0; k < od.GetNumberOfEmptyPath(); ++k) {
                    int j = od.GetEmptyPathIndexes()[k];
                    // Item 4: Demurrage of self-owned containers for empty path
                    expr += p_.GetEmptyPathDemurrageCost()[j] * z_var_[i][k];
                }
            }

            obj_ = IloMinimize(env_, expr);
            model_.add(obj_);
        }
        catch (const IloException& e) {
            std::cerr << "Error in SetObjectives: " << e.getMessage() << std::endl;
        }
    }

    void CapacityCalculate::SetConstraints()
    {
        SetConstraint1();
        SetConstraint2();
        SetConstraint3();
    }

    void CapacityCalculate::SetConstraint1() {
        try {
            c1_.resize(p_.GetDemand().size());

            // 对于每个 i ∈ I
            for (size_t i = 0; i < p_.GetDemand().size(); ++i) {
                IloExpr left(env_);

                const Request& od = in_.GetRequests()[i];
                // φ
                for (size_t k = 0; k < od.GetNumberOfLadenPath(); ++k) {
                    left += x_var_[i][k];
                    left += y_var_[i][k];
                }

                left += g_var_[i];

                std::string constr_name = "C1(" + std::to_string(i + 1) + ")";
                c1_[i] = IloRange(p_.GetDemand()[i] <= left <= p_.GetDemand()[i]);
                c1_[i].setName(constr_name.c_str());
                model_.add(c1_[i]);
            }
        }
        catch (const IloException& e) {
            std::cerr << "Error in SetConstraint1: " << e.getMessage() << std::endl;
        }
    }

    void CapacityCalculate::SetConstraint2() {
        try {
            // 对于每个 <n,n'> ∈ A'
            for (size_t nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn)
            {
                std::string constr_name = "C2(" + std::to_string(nn + 1) + ")";

                IloExpr left(env_);

                // 对于每个 i ∈ I
                for (size_t i = 0; i < p_.GetDemand().size(); ++i) {
                    Request od = in_.GetRequests()[i];

                    // φ
                    for (size_t k = 0; k < od.GetNumberOfLadenPath(); ++k) {
                        int j = od.GetLadenPathIndexes()[k];
                        left += p_.GetArcAndPath()[nn][j] * x_var_[i][k];
                        left += p_.GetArcAndPath()[nn][j] * y_var_[i][k];
                    }

                    //θ
                    for (size_t k = 0; k < od.GetNumberOfEmptyPath(); ++k) {
                        int j = od.GetEmptyPathIndexes()[k];
                        left += p_.GetArcAndPath()[nn][j] * z_var_[i][k];
                    }
                }

                //Vessel capacity
                for (size_t w = 0; w < p_.GetVesselPathSet().size(); ++w)
                {
                    int r = in_.GetVesselPathSet()[w].GetRouteID() - 1;
                    left -= p_.GetArcAndVesselPath()[nn][w] * p_.GetShipRouteAndVesselPath()[r][w] * v_var_[r][w];
                }

                IloRange c2 = IloRange(left <= 0);
                c2.setName(constr_name.c_str());
                model_.add(c2);
            }
        }
        catch (const IloException& e) {
            std::cerr << "Error in SetConstraint2: " << e.getMessage() << std::endl;
        }
    }


    void CapacityCalculate::SetConstraint3() {
        try {
            // 对于每个 p ∈ P
            for (size_t pp = 0; pp < p_.GetPortSet().size(); ++pp) {
                // 对于每个 t ∈ T
                for (size_t t = 1; t < p_.GetTimePointSet().size(); ++t) {
                    IloExpr left(env_);

                    // 对于每个 i ∈ I
                    for (size_t i = 0; i < p_.GetDemand().size(); ++i) {
                        const Request& od = in_.GetRequests()[i];

                        // 输入流 Z
                        if (p_.GetOriginOfDemand()[i] == p_.GetPortSet()[pp]) {
                            for (size_t k = 0; k < od.GetNumberOfEmptyPath(); ++k) {
                                int j = od.GetEmptyPathIndexes()[k];
                                for (size_t nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn) {
                                    if (in_.GetTravelArcs()[nn].GetDestinationPort() == p_.GetPortSet()[pp] &&
                                        in_.GetTravelArcs()[nn].GetDestinationTime() <= t &&
                                        in_.GetTravelArcs()[nn].GetDestinationTime() >= 1) {
                                        left += p_.GetArcAndPath()[nn][j] * z_var_[i][k];
                                    }
                                }
                            }
                        }

                        // 输入流 X
                        if (p_.GetDestinationOfDemand()[i] == p_.GetPortSet()[pp]) {
                            for (size_t k = 0; k < od.GetNumberOfLadenPath(); ++k) {
                                int j = od.GetLadenPathIndexes()[k];
                                for (size_t nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn) {
                                    if (in_.GetTravelArcs()[nn].GetDestinationPort() == p_.GetPortSet()[pp] &&
                                        in_.GetTravelArcs()[nn].GetDestinationTime() <= t - p_.GetTurnOverTimeSet()[pp] &&
                                        in_.GetTravelArcs()[nn].GetDestinationTime() >= 1) {
                                        left += p_.GetArcAndPath()[nn][j] * x_var_[i][k];
                                    }
                                }
                            }
                        }

                        // 输出流 X
                        if (p_.GetOriginOfDemand()[i] == p_.GetPortSet()[pp]) {
                            for (size_t k = 0; k < od.GetNumberOfLadenPath(); ++k) {
                                int j = od.GetLadenPathIndexes()[k];
                                for (size_t nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn) {
                                    if (in_.GetTravelArcs()[nn].GetOriginPort() == p_.GetPortSet()[pp] &&
                                        in_.GetTravelArcs()[nn].GetOriginTime() <= t &&
                                        in_.GetTravelArcs()[nn].GetOriginTime() >= 1) {
                                        left -= p_.GetArcAndPath()[nn][j] * x_var_[i][k];
                                    }
                                }
                            }
                        }

                        // 输出流 Z
                        for (size_t k = 0; k < od.GetNumberOfEmptyPath(); ++k) {
                            int j = od.GetEmptyPathIndexes()[k];
                            for (size_t nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn) {
                                if (in_.GetTravelArcs()[nn].GetOriginPort() == p_.GetPortSet()[pp] &&
                                    in_.GetTravelArcs()[nn].GetOriginTime() <= t &&
                                    in_.GetTravelArcs()[nn].GetOriginTime() >= 1) {
                                    left -= p_.GetArcAndPath()[nn][j] * z_var_[i][k];
                                }
                            }
                        }
                    }

                    std::string constr_name = "C3(" + std::to_string(pp + 1) + ")(" + std::to_string(t) + ")";
                    IloRange c3 = IloRange(left >= -p_.GetInitialEmptyContainer()[pp]);
                    c3.setName(constr_name.c_str());
                    model_.add(c3);
                }
            }
        }
        catch (const IloException& e) {
            std::cerr << "Error in SetConstraint3: " << e.getMessage() << std::endl;
        }
    }


    void CapacityCalculate::SolveModel() {
        try {
            if (cplex_.solve()) {
                // 可能的详细打印函数
                // PrintDetail();
                SetObjVal(cplex_.getObjValue());
                SetMinV();
                ChangeDemandVariation();
                if (cplex_.solve()) {
                    // 可能的详细打印函数
                    // PrintDetail();
                    SetMaxV();
                }
                else {
                    std::cout << "Exit with an error during re-solving" << std::endl;
                }
                PrintSolution();
            }
            else {
                std::cout << "No Solution" << std::endl;
            }
        }
        catch (const IloException& e) {
            std::cerr << "Concert Error: " << e.getMessage() << std::endl;
        }
    }


    void CapacityCalculate::ChangeDemandVariation() {
        try {
            for (size_t i = 0; i < p_.GetDemand().size(); ++i) {
                double newBound = p_.GetDemand()[i] + p_.GetMaximumDemandVariation()[i];
                c1_[i].setBounds(newBound, newBound);
            }
        }
        catch (const IloException& e) {
            std::cerr << "Error in ChangeDemandVariation: " << e.getMessage() << std::endl;
        }
    }


    void CapacityCalculate::SetMinV() {
        try {
            min_v_.resize(p_.GetVesselRouteSet().size(), std::vector<int>(p_.GetVesselPathSet().size()));

            for (size_t w = 0; w < p_.GetVesselPathSet().size(); ++w) {
                int r = in_.GetVesselPathSet()[w].GetRouteID() - 1;
                min_v_[r][w] = static_cast<int>(cplex_.getValue(v_var_[r][w]));
            }
        }
        catch (const IloException& e) {
            std::cerr << "Error in SetMinV: " << e.getMessage() << std::endl;
        }
    }


    void CapacityCalculate::SetMaxV() {
        try {
            max_v_.resize(p_.GetVesselRouteSet().size(), std::vector<int>(p_.GetVesselPathSet().size()));

            for (size_t w = 0; w < p_.GetVesselPathSet().size(); ++w) {
                int r = in_.GetVesselPathSet()[w].GetRouteID() - 1;
                max_v_[r][w] = static_cast<int>(cplex_.getValue(v_var_[r][w]));
            }
        }
        catch (const IloException& e) {
            std::cerr << "Error in SetMaxV: " << e.getMessage() << std::endl;
        }
    }

    void CapacityCalculate::PrintSolution() {
        std::cout << "Vessel Decision vVar : " << std::endl;
        for (size_t r = 0; r < p_.GetVesselRouteSet().size(); ++r) {
            std::cout << p_.GetVesselRouteSet()[r] << ":\t";
            for (size_t w = 0; w < p_.GetVesselPathSet().size(); ++w) {
                if (p_.GetShipRouteAndVesselPath()[r][w] != 0) {
                    std::cout << p_.GetVesselPathSet()[w];
                    std::cout << "(" << min_v_[r][w] << "~" << max_v_[r][w] << ")\t";
                }
            }
            std::cout << std::endl;
        }
        std::cout << std::endl;
    }

    void CapacityCalculate::PrintDetail() {
        try {
            for (size_t i = 0; i < p_.GetDemand().size(); ++i) {
                const Request& od = in_.GetRequests()[i];
                std::cout << "Demand" << (i + 1) << "("
                    << od.GetOriginPort() << "->" << od.GetDestinationPort() << ")"
                    << "(" << od.GetW_i_Earliest() << "->" << od.GetLatestDestinationTime() << ")"
                    << ":\t" << p_.GetDemand()[i] << " \t=\t";

                double totalX = 0;
                double totalY = 0;
                // φ
                for (size_t k = 0; k < od.GetNumberOfLadenPath(); ++k) {
                    totalX += cplex_.getValue(x_var_[i][k]);
                    totalY += cplex_.getValue(y_var_[i][k]);
                }

                std::cout << totalX << "\t\t" << totalY << "\t\t" << cplex_.getValue(g_var_[i]) << "\n";
            }
            std::cout << std::endl;
        }
        catch (const IloException& e) {
            std::cerr << "Error in PrintDetail: " << e.getMessage() << std::endl;
        }
    }


}

File: src/multi/model/CapacityCalculate.java
package multi.model;

import ilog.concert.*;
import ilog.cplex.IloCplex;
import multi.data.InputData;
import multi.data.Parameter;
import multi.structure.Request;

import java.util.ArrayList;

public class CapacityCalculate extends BasePrimalModel {
    public CapacityCalculate(InputData in, Parameter p) {
        super();
        this.in = in;
        this.p = p;
        this.ModelName = "CC" + "-R"+ in.getShipRouteSet().size() + "-T" + p.getTimeHorizon() + "-"+ FleetType + "-S" + randomSeed;
        try{
            cplex = new IloCplex();
            publicSetting(cplex);
            frame();

            solveModel();
        }catch (IloException e)
        {
            e.printStackTrace();
        }
    }

    public void setDecisionVars() throws IloException
    {
        String varName;

        SetVesselDecisionVars();

        vVar =new IloIntVar [p.getShippingRouteSet().length][p.getVesselPathSet().length];
        for (int w = 0; w < p.getVesselPathSet().length; w++) {
            int r = in.getVesselPathSet().get(w).getRouteID() - 1;
            varName = "v("+ (r+1)+")("+(w+1)+")";
            vVar[r][w] = cplex.boolVar(varName);
        }


        xVar = new ArrayList<>();
        yVar = new ArrayList<>();
        zVar = new ArrayList<>();

        gVar =new IloNumVar [p.getDemand().length];




        for(int i=0;i<p.getDemand().length;i++)
        {
            Request od = in.getRequestSet().get(i);

            IloNumVar[] xxxVar_k = new IloNumVar[od.getNumberOfLadenPath()];
            IloNumVar[] yyyVar_k = new IloNumVar[od.getNumberOfLadenPath()];
            IloNumVar[] zzzVar_k = new IloNumVar[od.getNumberOfEmptyPath()];

            xVar.add(xxxVar_k);
            yVar.add(yyyVar_k);
            zVar.add(zzzVar_k);

            for (int j = 0; j < od.getNumberOfLadenPath(); j++) {
                varName = "x(" + (i+1) +")";
                xVar.get(i)[j]=cplex.numVar(0,Integer.MAX_VALUE, varName);
                varName = "y(" + (i+1) +")";
                yVar.get(i)[j]=cplex.numVar(0,Integer.MAX_VALUE, varName);
            }
            for (int j = 0; j < od.getNumberOfEmptyPath(); j++) {
                varName = "z(" + (i+1) +")";
                zVar.get(i)[j]=cplex.numVar(0,Integer.MAX_VALUE, varName);
            }

            varName = "g(" + (i+1) +")";
            gVar[i]=cplex.numVar(0,Integer.MAX_VALUE, varName);
        }
    }

    public void setObjectives() throws IloException
    {
        IloLinearNumExpr Obj=cplex.linearNumExpr();

        // item1 : Operating Cost
        // r∈R
        for(int w = 0; w < p.getVesselPathSet().length; w++) {
            int r = in.getVesselPathSet().get(w).getRouteID() - 1;
            Obj.addTerm(p.getShipRouteAndVesselPath()[r][w]
                    , vVar[r][w]	);
        }

        // i
        for(int i=0;i<p.getDemand().length;i++)
        {
            // item2 : Penalty Cost of unsatisfied Demand : c2i*Gi
            Obj.addTerm(p.getPenaltyCostForDemand()[i], gVar[i]);

            Request od = in.getRequestSet().get(i);
            // \phi_i
            for(int k=0;k<od.getNumberOfLadenPath();k++)
            {
                int j = od.getLadenPathIndexes()[k];
                // item3 : Demurrage of self-owned and leased containers and Rental cost on laden paths
                Obj.addTerm(p.getLadenPathDemurrageCost()[j], xVar.get(i)[k]);
                Obj.addTerm(p.getLadenPathDemurrageCost()[j], yVar.get(i)[k]);
                Obj.addTerm(p.getRentalCost()*p.getTravelTimeOnPath()[j], yVar.get(i)[k]);
            }

            // item4: Demurrage of self-owned containers for empty path
            // \theta
            for(int k=0;k<od.getNumberOfEmptyPath();k++)
            {
                int j = od.getEmptyPathIndexes()[k];
                Obj.addTerm(p.getEmptyPathDemurrageCost()[j], zVar.get(i)[k]);
            }
        }

        cplex.addMinimize(Obj);
    }

    private IloRange[] C1;

    public void setConstraints() throws IloException
    {
        setConstraint1();
        setConstraint2();
        setConstraint3();
    }

     //(21)demand equation
    private void setConstraint1() throws IloException
    {
        C1 = new IloRange[p.getDemand().length];

        //∀i∈I
        for(int i=0;i<p.getDemand().length;i++)
        {
            IloLinearNumExpr left=cplex.linearNumExpr();

            Request od = in.getRequestSet().get(i);
            //φ
            for(int k=0;k<od.getNumberOfLadenPath();k++)
            {
                left.addTerm(1, xVar.get(i)[k]);
                left.addTerm(1, yVar.get(i)[k]);
            }

            left.addTerm(1, gVar[i]);

            String ConstrName = "C1("+(i+1)+")";
            C1[i] = cplex.addEq(left ,p.getDemand()[i], ConstrName);
        }
    }

    //(22) Vessel Capacity Constraint
    private void setConstraint2() throws IloException    {
        // ∀<n,n'>∈A'
        for(int nn = 0; nn<p.getTravelingArcsSet().length; nn++)
        {
            IloLinearNumExpr left=cplex.linearNumExpr();

            // i∈I
            for(int i=0;i<p.getDemand().length;i++)
            {
                Request od = in.getRequestSet().get(i);

                // φ
                for(int k=0; k<od.getNumberOfLadenPath(); k++)
                {
                    int j = od.getLadenPathIndexes()[k];

                    left.addTerm(p.getArcAndPath()[nn][j], xVar.get(i)[k]);
                    left.addTerm(p.getArcAndPath()[nn][j], yVar.get(i)[k]);
                }

                //θ
                for(int k=0;k<od.getNumberOfEmptyPath();k++)
                {
                    int j = od.getEmptyPathIndexes()[k];

                    left.addTerm(p.getArcAndPath()[nn][j], zVar.get(i)[k]);
                }
            }

            // vessel capacity
            // r∈R
            // ω∈Ω
            for(int w=0;w<p.getVesselPathSet().length;w++)
            {
                int r = in.getVesselPathSet().get(w).getRouteID() - 1;
                left.addTerm(-p.getArcAndVesselPath()[nn][w]
                                *p.getShipRouteAndVesselPath()[r][w]
                        , vVar[r][w]	);
            }
            String ConstrName = "C3"+"("+(nn+1)+")";
            cplex.addLe(left,0, ConstrName);
        }
    }

    // (24)Containers flow conservation
    private void setConstraint3() throws IloException
    {
        // p \in P
        for(int pp = 0; pp < p.getPortSet().length; pp++)
        {
            // t \in T
            for(int t = 1; t < p.getTimePointSet().length; t++)
            {
                IloLinearNumExpr left = cplex.linearNumExpr();

                // i \in I
                for(int i = 0; i < p.getDemand().length; i++)
                {
                    Request od = in.getRequestSet().get(i);

                    // Input Z flow:
                    // (item1)
                    // o(i) == p
                    if(p.getPortSet()[pp].equals(p.getOriginOfDemand()[i]))
                    {
                        //θi
                        for (int k = 0; k < od.getNumberOfEmptyPath(); k++)
                        {
                            int j = od.getEmptyPathIndexes()[k];

                            // <n,n'> ∈A'
                            for (int nn = 0; nn < p.getTravelingArcsSet().length; nn++) {
                                // p(n') == p
                                // 1<= t(n')<= t
                                if (in.getTravelingArcSet().get(nn).getDestinationPort().equals(p.getPortSet()[pp])
                                        && in.getTravelingArcSet().get(nn).getDestinationTime() <= t
                                        && in.getTravelingArcSet().get(nn).getDestinationTime() >= 1)
                                {
                                    left.addTerm(p.getArcAndPath()[nn][j], zVar.get(i)[k]);
                                }
                            }
                        }
                    }


                    // Input flow X
                    // item2
                    // d(i) == p
                    if(p.getPortSet()[pp].equals(p.getDestinationOfDemand()[i]))
                    {
                        for(int k=0;k<od.getNumberOfLadenPath();k++)
                        {
                            int j = od.getLadenPathIndexes()[k];

                            // <n,n'>∈A'
                            for(int nn = 0; nn<p.getTravelingArcsSet().length; nn++)
                            {
                                // p(n‘)∈p
                                // 1 <= t(n')<= t-sp
                                if(in.getTravelingArcSet().get(nn).getDestinationPort().equals(p.getPortSet()[pp])
                                        &&in.getTravelingArcSet().get(nn).getDestinationTime()<=t-p.getTurnOverTime()[pp]
                                        &&in.getTravelingArcSet().get(nn).getDestinationTime()>=1)
                                {
                                    left.addTerm(p.getArcAndPath()[nn][j], xVar.get(i)[k]);
                                }
                            }
                        }
                    }


                    //Output  flow X
                    // item3
                    // o(i) == p
                    if(p.getPortSet()[pp].equals(p.getOriginOfDemand()[i]))
                    {
                        // φi
                        for(int k=0;k<od.getNumberOfLadenPath();k++)
                        {
                            int j = od.getLadenPathIndexes()[k];

                            // <n.n'>∈A'
                            for(int nn = 0; nn<p.getTravelingArcsSet().length; nn++)
                            {
                                //p(n) == p
                                // t(n) <= t
                                if(in.getTravelingArcSet().get(nn).getOriginPort().equals(p.getPortSet()[pp])
                                        &&in.getTravelingArcSet().get(nn).getOriginTime() <=t
                                        &&in.getTravelingArcSet().get(nn).getOriginTime() >=1)
                                {
                                    left.addTerm(-p.getArcAndPath()[nn][j], xVar.get(i)[k]);
                                }
                            }
                        }
                    }


                    // Output Flow Z
                    // item4
                    // θ
                    for(int k=0;k<od.getNumberOfEmptyPath();k++)
                    {
                        int j = od.getEmptyPathIndexes()[k];

                        // <n,n'>∈A'
                        for(int nn = 0; nn<p.getTravelingArcsSet().length; nn++)
                        {
                            // p(n) == p
                            // t(n) <= t
                            if(in.getTravelingArcSet().get(nn).getOriginPort().equals(p.getPortSet()[pp])
                                    && in.getTravelingArcSet().get(nn).getOriginTime()<=t
                                    && in.getTravelingArcSet().get(nn).getOriginTime()>=1)
                            {
                                left.addTerm(-p.getArcAndPath()[nn][j], zVar.get(i)[k]);
                            }
                        }
                    }
                }
                String ConstrName = "C3("+(pp+1)+")("+(t)+")";
                cplex.addGe(left, -p.getInitialEmptyContainer()[pp], ConstrName);
            }
        }
    }

    public void solveModel()
    {
        try
        {
            if (WhetherExportModel)
                exportModel();
            if (cplex.solve())
            {
//                printDetail();
                setObjVal(cplex.getObjValue());
                setMinV();
                changeDemandVaration();
                if (cplex.solve())
                {
//                    printDetail();
                    setMaxV();
                }
                else
                {
                    System.out.println("Exit an error");
                }
                printSolution();
            }
            else
            {
                System.out.println("No Solution");
            }
        }
        catch (IloException ex)
        {
            System.out.println("Concert Error: " + ex);
        }
    }

    private void changeDemandVaration() throws IloException {
        for (int i = 0; i < p.getDemand().length; i++) {
            C1[i].setBounds(p.getDemand()[i]+p.getMaximumDemandVariation()[i]
                    ,p.getDemand()[i]+p.getMaximumDemandVariation()[i] );
        }
    }

    public void setMinV() throws IloException {
        minV = new  int[p.getShippingRouteSet().length][p.getVesselPathSet().length];
        for (int w = 0; w < p.getVesselPathSet().length; w++) {
            int r = in.getVesselPathSet().get(w).getRouteID() - 1;
            minV[r][w] =  (int) cplex.getValue(vVar[r][w]);
        }
    }
    public void setMaxV() throws IloException {
        maxV = new  int[p.getShippingRouteSet().length][p.getVesselPathSet().length];
        for (int w = 0; w < p.getVesselPathSet().length; w++) {
            int r = in.getVesselPathSet().get(w).getRouteID() - 1;
            maxV[r][w] =  (int) cplex.getValue(vVar[r][w]);
        }
    }
    private int[][] minV;
    private int[][] maxV;
    public void printSolution()    {
        System.out.println("Vessel Decision vVar : ");
        for(int r = 0; r<p.getShippingRouteSet().length; r++)
        {
            System.out.print(p.getShippingRouteSet()[r]+":\t");
            for (int w = 0; w < p.getVesselPathSet().length; w++) {
                if(p.getShipRouteAndVesselPath()[r][w] != 0)
                {
                    System.out.print(p.getVesselPathSet()[w]);
                    System.out.print("("+minV[r][w] + "~" + maxV[r][w] + ")\t");
                }
            }
            System.out.println();
        }
        System.out.println();
    }
    public void printDetail() throws IloException {
        for (int i = 0; i < p.getDemand().length; i++) {
            System.out.print("Demand"+(i+1)
                    +"("+in.getRequestSet().get(i).getOriginPort()
                    +"->"+in.getRequestSet().get(i).getDestinationPort()+")"
                    +"("+in.getRequestSet().get(i).getW_i_Earliest()
                    +"->"+in.getRequestSet().get(i).getLatestDestinationTime()+")"
                    +":\t"
                    +p.getDemand()[i]+" \t=\t");

            double totalX = 0;
            double totalY = 0;
            Request od = in.getRequestSet().get(i);
            //φ
            for(int k=0;k<od.getNumberOfLadenPath();k++)
            {
                totalX += cplex.getValue(xVar.get(i)[k]);
                totalY += cplex.getValue(yVar.get(i)[k]);
            }

            System.out.print(totalX+"\t\t");
            System.out.print(totalY+"\t\t");

            System.out.println(cplex.getValue(gVar[i]) + "\t");

        }
    }
}
